File: app/src/main/java/test/juyoufuli/com/myapplication/mvp/presenter/MainPresenter.java
package test.juyoufuli.com.myapplication.mvp.presenter;

import android.app.Application;
import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.OnLifecycleEvent;
import android.support.v4.app.Fragment;
import android.support.v4.app.SupportActivity;

import com.jess.arms.di.scope.FragmentScope;
import com.jess.arms.integration.AppManager;
import com.jess.arms.mvp.BasePresenter;
import com.jess.arms.utils.PermissionUtil;
import com.jess.arms.utils.RxLifecycleUtils;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import me.jessyan.rxerrorhandler.core.RxErrorHandler;
import me.jessyan.rxerrorhandler.handler.ErrorHandleSubscriber;
import me.jessyan.rxerrorhandler.handler.RetryWithDelay;
import test.juyoufuli.com.myapplication.mvp.entity.ArticleBean;
import test.juyoufuli.com.myapplication.mvp.entity.ArticleResponse;
import test.juyoufuli.com.myapplication.mvp.entity.BannerResponse;
import test.juyoufuli.com.myapplication.mvp.model.contract.MainContract;
import test.juyoufuli.com.myapplication.mvp.ui.home.adapter.ArticleAdapter;

/**
 * Author : ludf
 * Created Time : 2018-09-29  13:25
 * Description:
 */
@FragmentScope
public class MainPresenter extends BasePresenter<MainContract.Model, MainContract.View> {

    @Inject
    RxErrorHandler mErrorHandler;
    @Inject
    AppManager mAppManager;
    @Inject
    Application mApplication;
    @Inject
    List<ArticleBean> mUsers;
    @Inject
    ArticleAdapter mAdapter;

    private int lastUserId = 1;
    private boolean isFirst = true;
    private int preEndIndex;


    @Inject
    public MainPresenter(MainContract.Model model, MainContract.View rootView) {
        super(model, rootView);
    }

    /**
     * 使用 2017 Google IO 发布的 Architecture Components 中的 Lifecycles 的新特性 (此特性已被加入 Support library)
     * 使 {@code Presenter} 可以与 {@link SupportActivity} 和 {@link Fragment} 的部分生命周期绑定
     */
    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    void onCreate() {
//        requestUsers(true);//打开 App 时自动加载列表
//        requestBannerDataList();
    }

    public void requestUsers(final boolean pullToRefresh) {
        //请求外部存储权限用于适配android6.0的权限管理机制
        PermissionUtil.externalStorage(new PermissionUtil.RequestPermission() {
            @Override
            public void onRequestPermissionSuccess() {
                //request permission success, do something.
                requestFromModel(pullToRefresh);
            }

            @Override
            public void onRequestPermissionFailure(List<String> permissions) {
                mRootView.showMessage("Request permissions failure");
                mRootView.hideLoading();//隐藏下拉刷新的进度条
            }

            @Override
            public void onRequestPermissionFailureWithAskNeverAgain(List<String> permissions) {
                mRootView.showMessage("Need to go to the settings");
                mRootView.hideLoading();//隐藏下拉刷新的进度条
            }
        }, mRootView.getRxPermissions(), mErrorHandler);
    }

    private void requestFromModel(boolean pullToRefresh) {
        if (pullToRefresh) lastUserId = 1;//下拉刷新默认只请求第一页

        //关于RxCache缓存库的使用请参考 http://www.jianshu.com/p/b58ef6b0624b

        boolean isEvictCache = pullToRefresh;//是否驱逐缓存,为ture即不使用缓存,每次下拉刷新即需要最新数据,则不使用缓存

        if (pullToRefresh && isFirst) {//默认在第一次下拉刷新时使用缓存
            isFirst = false;
            isEvictCache = false;
        }

        mModel.getUsers(lastUserId, isEvictCache)
                .subscribeOn(Schedulers.io())
                .retryWhen(new RetryWithDelay(3, 2))//遇到错误时重试,第一个参数为重试几次,第二个参数为重试的间隔
                .doOnSubscribe(disposable -> {
                    if (pullToRefresh)
                        mRootView.showLoading();//显示下拉刷新的进度条
                    else
                        mRootView.startLoadMore();//显示上拉加载更多的进度条
                }).subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread())
                .doFinally(() -> {
                    if (pullToRefresh)
                        mRootView.hideLoading();//隐藏下拉刷新的进度条
                    else
                        mRootView.endLoadMore();//隐藏上拉加载更多的进度条
                })
                .compose(RxLifecycleUtils.bindToLifecycle(mRootView))//使用 Rxlifecycle,使 Disposable 和 Activity 一起销毁
                .subscribe(new ErrorHandleSubscriber<ArticleResponse>(mErrorHandler) {
                    @Override
                    public void onNext(ArticleResponse users) {
                        lastUserId = users.getData().getCurPage();//记录最后一个id,用于下一次请求
                        if (pullToRefresh) mUsers.clear();//如果是下拉刷新则清空列表
                        preEndIndex = mUsers.size();//更新之前列表总长度,用于确定加载更多的起始位置
                        mUsers.addAll(users.getData().getDatas());
                        if (pullToRefresh)
                            mAdapter.notifyDataSetChanged();
                        else
                            mAdapter.notifyItemRangeInserted(preEndIndex, users.getData().getDatas().size());
                    }
                });
    }


    public void requestBannerDataList() {
        mModel.getBanner()
                .subscribeOn(Schedulers.io())
                .retryWhen(new RetryWithDelay(3, 2))
                .subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread())
                .compose(RxLifecycleUtils.bindToLifecycle(mRootView))
                .subscribe(new ErrorHandleSubscriber<BannerResponse>(mErrorHandler) {

                    @Override
                    public void onNext(BannerResponse systemDataResponse) {
                        mRootView.updateBanner(systemDataResponse);

                    }
                });
    }


    @Override
    public void onDestroy() {
        super.onDestroy();
        this.mAdapter = null;
        this.mUsers = null;
        this.mErrorHandler = null;
        this.mAppManager = null;
        this.mApplication = null;
    }
}

File: app/src/main/java/test/juyoufuli/com/myapplication/mvp/presenter/MainPresenter.kt
package test.juyoufuli.com.myapplication.mvp.presenter

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.OnLifecycleEvent
import android.support.v4.app.Fragment
import android.support.v4.app.SupportActivity
import com.jess.arms.di.scope.FragmentScope
import com.jess.arms.mvp.BasePresenter
import com.jess.arms.utils.RxLifecycleUtils
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers
import me.jessyan.rxerrorhandler.core.RxErrorHandler
import me.jessyan.rxerrorhandler.handler.ErrorHandleSubscriber
import me.jessyan.rxerrorhandler.handler.RetryWithDelay
import test.juyoufuli.com.myapplication.app.utils.LogUtils
import test.juyoufuli.com.myapplication.mvp.entity.*
import test.juyoufuli.com.myapplication.mvp.model.contract.MainContract
import test.juyoufuli.com.myapplication.mvp.ui.home.adapter2.MainRecyclerViewAdapter
import java.util.*
import javax.inject.Inject
import kotlin.collections.ArrayList

/**
 * Author : ludf
 * Created Time : 2018-09-29  13:25
 * Description:
 */
@FragmentScope
class MainPresenter @Inject
constructor(model: MainContract.Model, rootView: MainContract.View) : BasePresenter<MainContract.Model, MainContract.View>(model, rootView) {
    @JvmField
    @Inject
    internal var mErrorHandler: RxErrorHandler? = null
    @JvmField
    @Inject
    internal var mUsers: ArrayList<ArticleBean>? = null
    @JvmField
    @Inject
    internal var mAdapter: MainRecyclerViewAdapter? = null
    internal var mTopArticle: ArrayList<ArticleBean> = ArrayList()

    private var lastUserId = 0
    private var preEndIndex: Int = 0

    /**
     * 使用 2017 Google IO 发布的 Architecture Components 中的 Lifecycles 的新特性 (此特性已被加入 Support library)
     * 使 `Presenter` 可以与 [SupportActivity] 和 [Fragment] 的部分生命周期绑定
     */
    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    internal fun onCreate() {
        //        requestUsers(true);//打开 App 时自动加载列表
        //        requestBannerDataList();


    }


    fun mergeArticle(pullToRefresh: Boolean) {
        var topArticle = mModel.getTopArticle()
        var users = mModel.getUsers(0, pullToRefresh)

        Observable.zip(topArticle, users,
                BiFunction<TopArticleResponse, ArticleResponse, ArrayList<ArticleBean>> { t1, t2 ->
                    val list = ArrayList<ArticleBean>()
                    list.addAll(t1.data)
                    list.addAll(t2.data.datas)
                    mTopArticle.clear()
                    mTopArticle.addAll(t1.data)
                    list
                }

        ).subscribeOn(Schedulers.io())
                .doOnSubscribe { disposable ->
                    if (pullToRefresh)
                        mRootView.showLoading()//显示下拉刷新的进度条
                    else
                        mRootView.startLoadMore()//显示上拉加载更多的进度条
                }
                .observeOn(AndroidSchedulers.mainThread())
                .doFinally {
                    if (pullToRefresh)
                        mRootView.hideLoading()//隐藏下拉刷新的进度条
                    else
                        mRootView.endLoadMore()//隐藏上拉加载更多的进度条
                }
                .compose(RxLifecycleUtils.bindToLifecycle<ArrayList<ArticleBean>>(mRootView))
                .subscribe(object : ErrorHandleSubscriber<ArrayList<ArticleBean>>(mErrorHandler!!) {
                    override fun onNext(t: ArrayList<ArticleBean>) {
                        LogUtils.d("全部文章合并：：" + t.size)
                        mUsers!!.clear()
                        mAdapter!!.mList!!.clear()

                        mUsers!!.addAll(t)
                        mAdapter!!.mList = mUsers
                        mAdapter!!.notifyDataSetChanged()
                        preEndIndex = mUsers!!.size
                    }
                })

    }

    fun requestTopArticle(firstTime: Boolean) {
        //请求外部存储权限用于适配android6.0的权限管理机制
        mModel.getTopArticle()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .compose(RxLifecycleUtils.bindToLifecycle<TopArticleResponse>(mRootView))
                .subscribe(object : ErrorHandleSubscriber<TopArticleResponse>(mErrorHandler!!) {
                    override fun onNext(t: TopArticleResponse) {
                        mTopArticle.clear()
                        mTopArticle.addAll(t.data)
                        if (mAdapter!!.itemCount != 0) {
                            mUsers!!.addAll(0, t.data)
                            mAdapter!!.addList(mUsers)
                            preEndIndex = mUsers!!.size
                        }
                    }
                })
    }

    fun requestFromModel(pullToRefresh: Boolean) {
        if (pullToRefresh) lastUserId = 0//下拉刷新默认只请求第一页
        //关于RxCache缓存库的使用请参考 http://www.jianshu.com/p/b58ef6b0624b
        val isEvictCache = true//是否驱逐缓存,为ture即不使用缓存,每次下拉刷新即需要最新数据,则不使用缓存
//        if (pullToRefresh && isFirst) {//默认在第一次下拉刷新时使用缓存
//            isFirst = false;
//            isEvictCache = false;
//        }
        mModel.getUsers(lastUserId, isEvictCache)
                .subscribeOn(Schedulers.io())
                .retryWhen(RetryWithDelay(3, 2))//遇到错误时重试,第一个参数为重试几次,第二个参数为重试的间隔
                .doOnSubscribe { disposable ->
                    if (pullToRefresh)
                        mRootView.showLoading()//显示下拉刷新的进度条
                    else
                        mRootView.startLoadMore()//显示上拉加载更多的进度条
                }.subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread())
                .doFinally {
                    if (pullToRefresh)
                        mRootView.hideLoading()//隐藏下拉刷新的进度条
                    else
                        mRootView.endLoadMore()//隐藏上拉加载更多的进度条
                }
                .compose(RxLifecycleUtils.bindToLifecycle<ArticleResponse>(mRootView))//使用 Rxlifecycle,使 Disposable 和 Activity 一起销毁
                .subscribe(object : ErrorHandleSubscriber<ArticleResponse>(mErrorHandler!!) {
                    override fun onNext(users: ArticleResponse) {
                        //下一页的页码
                        lastUserId = users.data.curPage
                        //如果是下拉刷新则清空列表
                        if (pullToRefresh) {
                            mUsers!!.clear()
                            mAdapter!!.clearList()
                        }
                        if (users.data.curPage != 1) {
                            mUsers!!.addAll(users.data.datas)
                            mAdapter!!.appendData(users.data.datas)
                        } else {
                            mUsers!!.clear()
                            mAdapter!!.clearList()
                            mUsers!!.addAll(0, mTopArticle)
                            mUsers!!.addAll(users.data.datas)

                            mAdapter!!.addList(mUsers)
                            LogUtils.d("变化后长度：：" + mUsers!!.size + "置顶的文章" + mTopArticle.size)
                        }
                        //更新之前列表总长度,用于确定加载更多的起始位置
                        preEndIndex = mUsers!!.size
                        if (pullToRefresh)
                            mAdapter!!.notifyDataSetChanged()
                        else
                            mAdapter!!.notifyItemRangeInserted(preEndIndex, users.data.datas.size)
                    }
                })
    }


    fun requestBannerDataList() {
        mModel.getBanner()
                .subscribeOn(Schedulers.io())
                .retryWhen(RetryWithDelay(3, 2))
                .observeOn(AndroidSchedulers.mainThread())
                .compose(RxLifecycleUtils.bindToLifecycle<BannerResponse>(mRootView))
                .subscribe(object : ErrorHandleSubscriber<BannerResponse>(mErrorHandler!!) {

                    override fun onNext(systemDataResponse: BannerResponse) {
                        mRootView.updateBanner(systemDataResponse)

                    }
                })
    }


    fun collectArticle(id: String) {
        mModel.collectArticle(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .compose(RxLifecycleUtils.bindToLifecycle<LoginResponse>(mRootView))
                .subscribe(object : ErrorHandleSubscriber<LoginResponse>(mErrorHandler!!) {

                    override fun onNext(systemDataResponse: LoginResponse) {

                    }
                })
    }


    fun cancelCollectArticle(id: String) {
        mModel.cancelCollectArticle(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .compose(RxLifecycleUtils.bindToLifecycle<LoginResponse>(mRootView))
                .subscribe(object : ErrorHandleSubscriber<LoginResponse>(mErrorHandler!!) {

                    override fun onNext(systemDataResponse: LoginResponse) {

                    }
                })
    }


    override fun onDestroy() {
        super.onDestroy()
        this.mAdapter = null
        this.mUsers = null
        this.mErrorHandler = null
    }
}
